src/main/java/de/tum/in/www1/artemis/service/plagiarism/ContinuousPlagiarismControlResultsService.java

```diff
@@ -0,0 +1,102 @@
+        this.submissionRepository = submissionRepository;
+        this.resultRepository = resultRepository;
+    }
+
+    void handleCpcResult(PlagiarismResult<?> result) {
+        var submissionsIdsWithPlagiarism = result.getComparisons().stream().flatMap(it -> Stream.of(it.getSubmissionA().getSubmissionId(), it.getSubmissionB().getSubmissionId()))
+                .collect(toUnmodifiableSet());
+
+        result.getExercise().getStudentParticipations().stream().filter(participation -> participation.findLatestSubmission().isPresent()).forEach(participation -> {
+            var submission = participation.findLatestSubmission().get();
+
+            boolean plagiarismDetected = submissionsIdsWithPlagiarism.contains(submission.getId());
+            submissionRepository.setPlagiarismDetected(plagiarismDetected, submission.getId());
+
+            if (plagiarismDetected) {
+                addResultWithPlagiarismFeedback(submission, participation);
+            }
+            else {
+                deletePastResultsWithPlagiarismFeedback(submission.getId());
+            }
+        });
+    }
+
+    private void addResultWithPlagiarismFeedback(Submission submission, StudentParticipation participation) {
+        log.debug("Adding cpc results for submission: submissionId={}.", submission.getId());
+
+        var result = new Result();
+        result.setSubmission(submission);
+        result.setAssessmentType(AssessmentType.AUTOMATIC);
+        result.setCompletionDate(ZonedDateTime.now());
+        result.setParticipation(participation);
+        result.setRated(true);
+        result.setExampleResult(false);
+        result.setSuccessful(false);
+        result.setScore(0.0, participation.getExercise().getCourseViaExerciseGroupOrCourseMember());
+
+        var feedback = new Feedback();
+        feedback.setType(FeedbackType.AUTOMATIC);
+        feedback.setPositive(false);
+        feedback.setResult(result);
+        feedback.setVisibility(Visibility.ALWAYS);
+        feedback.setCredits(0.0);
+
+        var feedbackText = participation.getStudent().map(User::getLangKey).map(Locale::forLanguageTag).filter(isEqual(Locale.ENGLISH)).map(it -> PLAGIARISM_RESULT_FEEDBACK_EN)
+                .orElse(PLAGIARISM_RESULT_FEEDBACK_DE);
+        feedback.setText(feedbackText);
+
+        result.setFeedbacks(List.of(feedback));
+
+        resultRepository.save(result);
+    }
+
+    private void deletePastResultsWithPlagiarismFeedback(long submissionId) {
+        log.debug("Removing cpc results for submission: submissionId={}.", submissionId);
+        resultRepository.findAllWithFeedbackBySubmissionId(submissionId).stream()
+                .filter(result -> result.getFeedbacks().stream()
+                        .anyMatch(it -> it.getText().contains(PLAGIARISM_RESULT_FEEDBACK_EN) || it.getText().contains(PLAGIARISM_RESULT_FEEDBACK_DE)))
```

Comment:
Extract the functionality into a new method.

Reply:
https://github.com/ls1intum/Artemis/pull/6804#discussion_r1264676868